keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2397 @@
+ 
+
+##### `NodeUnprepareResource`
+
+A Kubelet Plugin MUST implement this RPC call. This RPC is a reverse
+operation of `NodePrepareResource`. This RPC MUST undo the work by
+the corresponding `NodePrepareResource`. This RPC SHALL be called by
+kubelet at least once for each successful `NodePrepareResource`. The
+Plugin SHALL assume that this RPC will be executed on the node where
+the resource is being used.
+
+This RPC is called by kubelet when the Pod using the resource is being
+deleted.
+
+This operation MUST be idempotent. If this RPC failed, or kubelet does
+not know if it failed or not, it can choose to call
+`NodeUnprepareResource` again.
+
+```protobuf
+message NodeUnprepareResourceRequest {
+  // The ResourceClaim namespace (ResourceClaim.meta.Namespace).
+  // This field is REQUIRED.
+  string namespace = 1;
+  // The UID of the Resource claim (ResourceClaim.meta.UUID).
+  // This field is REQUIRED.
+  string claim_uid = 2;
+  // The name of the Resource claim (ResourceClaim.meta.Name)
+  // This field is REQUIRED.
+  string claim_name = 3;
+  // List of fully qualified CDI device names
+  // Kubelet plugin returns them in the NodePrepareResourceResponse
+  repeated string cdi_device = 4;
+}
+
+message NodeUnprepareResourceResponse {
+  // Intentionally empty.
+}
+```
+
+###### NodeUnprepareResource Errors
+
+If the plugin is unable to complete the NodeUprepareResource call
+successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST
+return the specified gRPC error code.  Kubelet MUST implement the
+specified error recovery behavior when it encounters the gRPC error
+code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+
+#### Implementing optional resources
+
+This can be handled entirely by a resource driver: its parameters can support a
+range starting at zero or a boolean flag that indicates that something is not a
+hard requirement. When asked to filter nodes for delayed allocation, the driver
+reports nodes where the resource is available and only falls back to those
+without it when resources are exhausted. When asked to allocate, it reserves
+actual resources if possible, but also proceeds with marking the ResourceClaim
+as allocated if that is not possible. Kubernetes then can schedule a pod using
+the ResourceClaim. The pod needs to determine through information passed in by
+the resource driver which resources are actually available to it.
+
+#### Implementing a plugin for node resources
+
+The proposal depends on a central resource driver controller. Implementing that
```

Comment:
I'm curious to get your thoughts... For some classes of resources today, i.e. GPUs with existing "simple" device plugin model, we don't need a central resource controller in the first place, and I think that's an advantage - at least from the simplicity standpoint :). 

A daemonset can install the GPU drivers and gpu device plugin, and the kubelet can report the device resources and there does not need to be any external communication to central controller, so we don't need to worry about setting up communication path between the node/device plugin/api-server.

Do we know what set of devices we need the central resource controller for? Is it specifically for cases of devices where they can be provisioned on the fly and not bound to the "life" of a node? Just trying to understand tradeoffs of this model vs extending device plugins as mentioned below.

Reply:
Let's say the node is a cloud resource that can have a local-enough GPU magically appear from an API call, and then disappear with another API call. I can see how it'd be nice to only attach that GPU when there is a Pod that can't be scheduled because it's waiting for a suitable node.

However, that's different from the user stories this KEP lists. Do any of the _listed_ user stories ned that central control?

Reply:
> Do we know what set of devices we need the central resource controller for?

The main question is whether the attribute of the device fits into a single integer. That is what kubelet knows how to communicate on behalf of the local device plugin. Anything more complex becomes problematic. @klueska provided a very good summary of how the current model is problematic for NVIDIA [in one of the comment on use cases above](https://github.com/kubernetes/enhancements/pull/3064#discussion_r895540324).

> Is it specifically for cases of devices where they can be provisioned on the fly and not bound to the "life" of a node?

Lifecycle may be another important aspect, with reuse and sharing not being supported by the device plugin API. It's not so much the life of a node, though.

Reply:
> However, that's different from the user stories this KEP lists. Do any of the listed user stories ned that central control?

We don't have user stories for all of the things that could be done with this new API. But that doesn't mean that they are out of scope or not important. The "Motivation" section has one point about "Support Over the Fabric devices".

Reply:
I know it'll extend the KEP, but I think it would be good to list at least one user story where that central / not-node-local control aspect is relevant.

Reply:
Makes sense. I'm currently holding off adding more commits to my branch, but locally I have added:

---- 

#### Network-attached accelerator

As a data center operator, I want to place accelerators in a separate enclosure
where they can get connected to different compute nodes through [Compute
Express Link™ (CXL™)](https://www.computeexpresslink.org/). This is more
cost-effective because not all workloads need these accelerators and more
flexible because worker nodes can be added or removed independently from
accelerators.

Jobs then compete for these accelerators by requesting them through a
ResourceClaim. The kube-scheduler ensures that all required resources for a pod
are reserved for it (CPU, RAM, and the ResourceClaim) before scheduling the pod
onto a worker node. The resource driver then dynamically attaches the
accelerator and detaches it again when the pod completes.

----

This covers two things that device plugin based approaches don't achieve:
- Guaranteed reservation of resources before scheduling (with Akri, pods may get scheduled and then get stuck while waiting for an IP camera to be released).
- Support for submitting more work than the cluster can handle (Akri is based on the assumption that users know how many IP cameras are available and avoid pods getting stuck by not submitting too many pods).

Reply:
So adding the CXL user story was not enough? The problem is that real devices for that are still in development. Given how complicated support for network attached devices will be, we need to start ASAP to work on that support. Starting when there is a real device will be too late. We can delay beta until there is one and validate our choices based on that, but shouldn't delay alpha.

Reply:
I could add something for IP cameras.

Reply:
I found the CXL case to be pretty demonstrative - it is not controllable from any node, but it is used from sepcific nodes.  It's the same problem as volume mounting (which is not surprising as this whole model draws on volumes as a reference).

Reply:
Anything accessed over network is less compelling to me - there isn't that aspect of "only certain nodes can access it"?

Reply:
> So adding the CXL user story was not enough? The problem is that real devices for that are still in development. Given how complicated support for network attached devices will be, we need to start ASAP to work on that support. 

@pohly to clarify, I think the CXL story is great example and suggest to add it into the KEP as a concrete use case. Regarding IP camera usecase, it also makes sense, but I personally would be very curious how [akri](https://github.com/project-akri/akri) handles it (they mention IP cameras as one of their usecases) since that project is completely out of tree and hacks _ontop_ of the device plugin api.